### DISCUSSION TITLE
Connections not being available at schema's type fields

### DISCUSSION THREAD

**Discussion by mvoto** [rmosolgo/graphql-ruby#4615] [Q&A] [answered]
Hello there, I am working with a federated GraphQL API and I need to add Apollo directives(using the [apollo's federation ruby gem](https://github.com/Gusto/apollo-federation-ruby)) to some specific connections and their fields(as they are auto-generated by graphql-ruby gem).
I am trying to loop and manipulate some connections from my schema objects, so I can add directives to specific objects, depending on certain conditions.

As an example, I have the following implementation:
```
# ./schema.rb
UnifiedGraph::Schema < GraphQL::Schema
  include ApolloFederation::Schema

  # ...

  class Tagging
    def self.find_contract_directive(field_definition)
      field_definition.directives.find do |directive|
        UnifiedGraph::Constants::PERMITTED_VERSIONS.include?(directive.arguments[:name].to_sym)
      end
    end

    def self.add_version_tag(field, version)
      field.extend(UnifiedGraph::Members::HasVersion)
      field.version version
    end
  end

  class ConnectionVersionTag < Tagging
    class << self
      def add(schema_types)
        schema_types.each do |type_name, klass|
          next unless klass.ancestors.include?(GraphQL::Schema::Object)

          klass.fields.each do |_field_name, field_definition|
            next unless field_definition.connection?
            contract_directive = find_contract_directive(field_definition)
            next unless contract_directive.present?

            field_definition.type.unwrap.fields.each do |_name, field|
              # Adds tags to pageInfo, edges, and nodes
              add_version_tag(field, contract_directive.arguments[:name].to_sym)
              # Adds tags to the children of the above
              add_version_tag_to_children(field, contract_directive.arguments[:name].to_sym)
            end
          end
        end
      end

      private

        def add_version_tag_to_children(parent_field, version)
          field_type = parent_field.type.unwrap
          # We should not dynamically apply tags to children of nodes,
          # because it applies the tag on the object. Instead, we should
          # explicitly apply tags on the fields of the object we want to expose.
          return if parent_field.name == 'nodes'

          add_version_tag(field_type, version)
        end
    end
  end

  # ...
```
Just had that implementation, so we could call `ConnectionVersionTag.add(types)` in other schema files that it inherits, ie: `ConnectionVersionTag.add(types)`.
This is the way we determine some connections:

```
field :bundles, Public::Graph::Bundle::Types::Bundle.connection_type, resolver: Public::Graph::Bundle::Resolvers::Bundles, description: "A list of the Site's Bundles", null: true, version: :beta
# ...
field :users, Public::Graph::User::Types::User.connection_type, resolver: Public::Graph::User::Resolvers::Users, description: "A list of the Site's Users", null: true, version: :beta
```
The above connections have different behavior when I loop through schema's `types`. The `bundles` connection is not being listed, while the `users` connection is. What could be the reason for this to happen? What could differentiate the connections(as they are defined the exact same way)?
Please, let me know if you need more details or clarification.

Currently using: 
- graphql-ruby on **1.12.24**
- apollo-federation on **3.6.5**

**Comment by rmosolgo** [comment]
Hey! I'm not sure what would explain the difference but my first guess is, do the `resolver:`s for those fields have `type ...` configurations? Maybe those are different between the two cases. 

Does the `bundles` field show up when you inspect the schema directly? For example, what do these lines return if you run them in a Rails console:

```ruby 
MySchema.get_type("Site").fields.keys

# Somehow reference your `Site` type class, however it's defined: 
Types::Site.fields.keys

bundles_field = Types::Site.get_field("bundles")
bundles_field.type.to_type_signature 
bundles_field.connection? 
```

**Reply by mvoto** [reply]
> MySchema.get_type("Site").fields.keys
From this comment, our team just noticed that the Site type(`Public::Graph::Site::Types`) is not available from our schema(which explains the custom logic not being able to grab the connections from the loop).
However, other fields are being properly tagged, even the connections, they are there, but without some fields(that the custom logic for tagging takes care of).

> Does the bundles field show up when you inspect the schema directly?
Yes, it does:
```
irb(main):013:0> Public::Graph::Site::Types::Site.get_field("bundles")
=> #<UnifiedGraph::Types::BaseField Site.bundles(...): BundleConnection>
irb(main):014:0> bundles_field = Public::Graph::Site::Types::Site.get_field("bundles")
=> #<UnifiedGraph::Types::BaseField Site.bundles(...): BundleConnection>
irb(main):015:0> bundles_field.type.to_type_signature
=> "BundleConnection"
irb(main):016:0> bundles_field.connection?
=> true
```

We also identified that the bundles field is not being properly tagged, because is not being declared anywhere else and the users field is being added because it is defined within another type(not only from Site).
Now we are not able to understand why we have field types available on schema and Site, for example, is not:

```
irb(main):001:0> Public::Graph::ThinkificSchema.get_type('User')
2023-09-06 19:39:13 +0000, WARN: Creating scope :search. Overwriting existing method Product.search. from 
=> Public::Graph::User::Types::User
irb(main):002:0> Public::Graph::ThinkificSchema.get_type('Bundle')
=> Public::Graph::Bundle::Types::Bundle
irb(main):003:0> Public::Graph::ThinkificSchema.get_type('Site')
=> nil
```

This is how Site is defined:

```
module Public::Graph::Site::Types
  class Site < Public::Graph::Types::BaseObject
  
    description 'Site'
    permitted_scopes []
    key fields: :id

   # ...
   field :bundles, Public::Graph::Bundle::Types::Bundle.connection_type, resolver: Public::Graph::Bundle::Resolvers::Bundles, description: "A list of the Site's Bundles", null: true, version: :beta
   field :users, Public::Graph::User::Types::User.connection_type, resolver: Public::Graph::User::Resolvers::Users, description: "A list of the Site's Users", null: true, version: :beta
   # ...
end
```

It's being added at the query type:

```
class Public::Graph::QueryType < Public::Graph::Types::BaseObject
  description 'Query root for the thinkific subgraph'

  permitted_scopes []
 
  # ...
  field :site, ::Public::Graph::Site::Types::Site, 'Site', null: true, version: :beta
  # ...
end
```

And at the schema file, we have something like:

```
class Public::Graph::ThinkificSchema < UnifiedGraph::Schema
  federation version: '2.0'

  default_max_page_size 100

  use BatchLoader::GraphQL

  query(Public::Graph::QueryType)
  mutation(Public::Graph::MutationType)

  # ...

end
```

We are now trying to understand why Site is not being considered as part of the schema.
Is there anything that seems weird or incorrect from the code that I just shared?

**Comment by mvoto** [comment, marked as answer]
Hey @rmosolgo it happens that the Site type wasn't available in the custom logic to add tags to specific connection fields, because it was not in schema's `types`, but it is defined within QueryType as mentioned above, so it should be handled differently.
From Schema, it can be accessed by `query.fields` instead.
After implementing that, we were able to manipulate and add tags to the connections fields of that specific Site type.

Thanks for the help! Guess we figured this one out.
